Delete product command

Criss-Wang opened the pull request:
Some room for improvement (feel free to give advice on these potential issues):

* Handle delete supplier/warehouse and their product in one method.
* The command is currently `delete TYPE INDEX pd/PRODUCT_NAME`, not sure if any bug may happen about this format.
* Define a new `getProductByName()` method in Warehouse and Supplier (No test written yet), not sure if it is desirable.
* Current deletion is just the creation of an updated supplier/warehouse object.

Criss-Wang committed:
Implement delete product function for CLI-nic

Criss-Wang committed:
Update test for DeleteCommandParser and userGuide

Criss-Wang committed:
Remove error data in JsonClinicStorageTest folder

There is an .idea folder created accidentally.

In light of this, let's delete this folder directly.

zhengweii commented on docs/UserGuide.md:
```diff
@@ -144,6 +144,22 @@ Examples:
+* `delete warehouse 1 pd/Panadol` Removes product with the name _Panadol_ from
+the warehouse at index 1 of the list of warehouse.
+* `delete supplier 12 pd/Aspirin` Removes product with the name _Aspirin_
+the from supplier at index 12 from the list of suppliers.
```
Minor typo here! Should swap "the" and "from"!

zhengweii replied:
Perhaps we should try to make the explanation consistent for both suppliers and warehouses??

Criss-Wang replied:
Thanks for spotting the error!

Criss-Wang replied:
> ld try to make the explana

May I know what do you mean by making the explanation consistent?

zhengweii replied:
Ahh oops! I was wondering if it would be possible to make the explanation for the examples in line 159 and 161 consistent? In line 159, it says "of the list of warehouse" and then in line 161 it says "from list of suppliers". Would it be possible to use the same format for both??

Criss-Wang replied:
I see, that's actually a typo also haha, thanks for pointing it out!

zhengweii commented on src/main/java/seedu/clinic/commons/core/Messages.java:
```diff
@@ -12,6 +12,10 @@
+    public static final String MESSAGE_INVALID_PRODUCT_NAME_IN_SUPPLIER =
+            "The product: %1$s is not sold by supplier: %2$s";
+    public static final String MESSAGE_INVALID_PRODUCT_NAME_IN_WAREHOUSE =
+            "The product: %1$s is not sold by warehouse: %2$s";
```
I think the message should be "not stored / contained in warehouse"!

Criss-Wang replied:
Thanks for the suggestion, your are right!

zhengweii commented on src/main/java/seedu/clinic/logic/commands/DeleteCommand.java:
```diff
@@ -21,59 +29,127 @@
             + "Example: " + COMMAND_WORD + " supplier 1\n"
-            + COMMAND_WORD + " warehouse 1";
+            + COMMAND_WORD + " warehouse 1\n\n"
+            + COMMAND_WORD
```
Just wondering if this line is extraneous?

Criss-Wang replied:
Actually I'm not sure if it is desirable format, maybe we can discuss after everything is done with their version of Message so we unify .

zhengweii commented on src/main/java/seedu/clinic/logic/commands/DeleteCommand.java:
```diff
@@ -21,59 +29,127 @@
-     * Parses the given {@code String} of arguments in the context of the DeleteCommand
-     * and returns a DeleteCommand object for execution.
-     * @throws ParseException if the user input does not conform the expected format.
+     * Creates an DeleteCommand to delete the warehouse/supplier at {@code targetIndex} of the displayed list.
```
I think it should be {@Index} instead of @code targetIndex! I believe they used the class instead of the argument name!

Criss-Wang replied:
This is just a styling for the parameter name targetIndex so I suppose it's okay

zhengweii commented on src/main/java/seedu/clinic/logic/commands/DeleteCommand.java:
```diff
@@ -21,59 +29,127 @@
+    }
+
+    /**
+     * Creates an DeleteCommand to delete the specified {@code targetProductName} in the warehouse/supplier
```
If we are using the class name instead of the argument name, then this part needs to be changed as well!

Criss-Wang replied:
I think in AB3 they mean argument name. So this part is probably okay

zhengweii replied:
Ahh okok! Thanks for clarifying :)

zhengweii commented on src/main/java/seedu/clinic/logic/commands/DeleteCommand.java:
```diff
@@ -21,59 +29,127 @@
-            return new CommandResult(String.format(MESSAGE_DELETE_SUPPLIER_SUCCESS, supplierToDelete));
+            return executeSupplierRelatedDeletion(model);
         } else {
-            List<Warehouse> lastShownList = model.getFilteredWarehouseList();
+            return executeWarehouseRelatedDeletion(model);
+        }
```
Perhaps we could just return the function without the else clause here!

Criss-Wang replied:
Thanks for the suggestion, will take that into account!

zhengweii commented on src/main/java/seedu/clinic/model/supplier/Supplier.java:
```diff
@@ -64,6 +65,21 @@ public Remark getRemark() {
     }
 
+    /**
+     * Returns a product with the {@code targetName} sold by the supplier.
```
This @code too if we are using class name!

zhengweii commented on src/main/java/seedu/clinic/model/warehouse/Warehouse.java:
```diff
@@ -69,6 +70,21 @@ public Address getAddress() {
     }
 
+    /**
+     * Returns a product with the {@code targetName} stored in the warehouse.
```
This too!

zhengweii commented on src/test/java/seedu/clinic/logic/commands/DeleteCommandTest.java:
```diff
@@ -79,6 +103,19 @@ public void execute_invalidIndexUnfilteredList_throwsCommandException() {
+
+        assertCommandFailure(deleteCommand, modelForSupplier, expectedMessage);
+    }
+
```
Do we have tests for invalid index when deleting products??

Criss-Wang replied:
I think this is covered in the delete supplier/warehouse part, I can add this test again later if needed, thanks for the comment!

zhengweii replied:
Ahh ok! Thanks for the update!!

Criss-Wang committed:
Update delete implementation

Criss-Wang committed:
Update delete command format

The group decides to standardize the command format with prefix-attached
style. The existing format need to be adjusted to make the right change.

Criss-Wang committed:
Fix styling issue and implement tests for the new format

Criss-Wang committed:
Merge branch 'master' into delete-product-command

tohyuting reviewed (approved):
LGTM, just some overlap with "Update DG for implementation of DeleteCommand". Do merge this branch to master before merging the mentioned branch as the DG for this branch may not be updated.
